src/ngraph_encapsulate_op.cc

```diff
@@ -165,51 +185,82 @@ class NGraphEncapsulateOp : public tf::OpKernel {
     // Allocate tensors for arguments.
     vector<shared_ptr<ng::runtime::TensorView>> ng_inputs;
 
-    auto& last_used_src_ptrs = m_last_used_src_ptrs_map[ng_function];
-    last_used_src_ptrs.resize(input_shapes.size());
+    std::vector<std::pair<void*, std::shared_ptr<ng::runtime::TensorView>>>&
+        input_caches = m_ng_function_input_cache_map[ng_function];
+    input_caches.resize(input_shapes.size());
 
```

Comment:
We should be able to check if it's cached and goes to copy directly for GPU and NNP if it is.

Reply:
I don't understand what you mean here...

Reply:
I think what @fengleitian means is that we can check that if all cached TensorViews are not `nullptr`, we can skip the shape and element type computation logic below, and go directly to CPU tensor freshness check or Device memory copy. The saving would be
```cpp
ng::Shape ng_shape(input_shapes[i].dims());
for (int j = 0; j < input_shapes[i].dims(); ++j) {
  ng_shape[j] = input_shapes[i].dim_size(j);
}
ng::element::Type ng_element_type;
OP_REQUIRES_OK(ctx, TFDataTypeToNGraphElementType(ctx->input(i).dtype(), &ng_element_type));
```
and the cost would be checking that all cached TensorViews are not `nullptr`

It seems like that would add additional complexity to the logic, while the saving is minimal.

Reply:
it will be like.
```
if(!cached)
{
  if(CPU)
  {
      create_tensorA();
  }
  else
  {
     create_tensorB();
  }
}
else
{
  if(CPU)
  {
      if(currnt_ptr != last_ptr)  create_tensorA();
  }
}

if(!CPU)  
{
   copy;
}
```
I'm not going to hold this PR for this.